### PR TITLE
feat: added support for pnpm lockfiles v9

### PR DESCRIPTION
**Pull request by gemaxim**
### What this does

Extended support for pnpm v9 lockfiles. Added tests to match.

### CODE REVIEW

File: lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts [hunk 1]
```diff
@@ -55,28 +55,30 @@ export abstract class PnpmLockfileParser {
...
-          optionalDependencies: versionData.optionalDependencies,
+          dependencies: versionData.dependencies || {},
+          devDependencies: versionData.devDependencies || {},
+          optionalDependencies: versionData.optionalDependencies || {},
         };
         packages[`${pkg.name}@${pkg.version}`] = pkg;
       },
     );
-    return packages;
+    this.extractedPackages = packages;
```
Comment (dotkas) [inline lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts:81]: Why are we doing it like this? Functions with side-effects looks a bit bad to me. But maybe I don't understand it.

**Review by dotkas — APPROVED** [review]
LGTM except I don't get the impure function change.

File: lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts (hunk 1 above)
Comment (gemaxim) [reply lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts:81]: becasue the new lockfile version specifies dependencies and other important data in the 'snapshots' field that's being populated in the subclass constructor, after calling the super constructor. But yes, it's better to return the value and assign it where I call the function.

Comment (gemaxim) [reply lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts]: fixed

**Pull request merged**